File: 206-reverse-linked-list/go/main.go
package main

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reverseList(head *ListNode) *ListNode {
	if head == nil || (*head).Next == nil {
		return head
	}

	a := head
	b := (*head).Next
	c := (*b).Next

	(*a).Next = nil

	for b != nil {
		(*b).Next = a
		a = b
		b = c
		if c != nil {
			c = (*c).Next
		}
	}

	return a
}

File: 206-reverse-linked-list/cpp/main.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
 public:
  ListNode* reverseList(ListNode* head) {
    if (head == NULL || head->next == NULL) {
      return head;
    }
    auto a = head;
    auto b = head->next;
    auto c = head->next->next;
    a->next = NULL;

    while (b != NULL) {
      b->next = a;
      a = b;
      b = c;
      if (c != NULL) {
        c = c->next;
      }
    }
    return a;
  }
};